lib/protocol/collection/notes_collection.ts

```diff
@@ -0,0 +1,94 @@
+import { SNNote, SNTag, SNSmartTag, SNPredicate, SNItem, ContentType } from '../../models'
+import { ItemCollection, CollectionSort, SortDirection } from "./item_collection";
+import { UuidString } from "@Lib/types";
+
+export class NotesCollection {
+  private displayedList: SNNote[] = [];
+  private tag?: SNTag;
+  needsRebuilding = true;
+
+  constructor(private collection: ItemCollection) {
+  }
+
+  private tags() {
+    return (this.collection.typedMap[ContentType.Tag] || []) as SNTag[];
+  }
+
+  public notesMatchingSmartTag(smartTag: SNSmartTag, notes: SNNote[]): SNNote[] {
+    let predicate = smartTag.predicate;
+
+    /** Optimized special cases */
+    if (smartTag.isArchiveTag) {
+      return notes.filter(note => note.archived && !note.trashed && !note.deleted);
+    } else if (smartTag.isTrashTag) {
+      return notes.filter(note => note.trashed && !note.deleted);
+    }
+    let matchingNotes: Pick<SNNote, 'uuid'>[] = notes.filter(
+      note => !note.trashed && !note.deleted
+    );
+    if (smartTag.isAllTag) {
+      return matchingNotes as SNNote[];
+    }
+
+    if (predicate.keypathIncludesVerb('tags')) {
+      /** Populate notes with their tags */
+      const tags = this.tags();
+      matchingNotes = notes.map(note => ({
+        ...note,
+        ...note.payload,
+        tags: tags.filter(tag => tag.hasRelationshipWithItem(note))
```

Comment:
1. What exactly is `tags` used for? After this condition you end up returning `this.collection.map[note.uuid]` and so the custom object you created here appears to be ignored?

2. What exactly is the type of `{...note, ...note.payload}`? Seems like an awkward concatenation. What is the purpose of combining them?

Reply:
1. `tags` is there so that existing predicates like `"tags.length"` can work.

2. Because I'm mapping to a plain object we lose every note getters, so I put the payload properties at the top-level to support other predicates. 
But you're right, it is kind of awkward. I'll get back to you with another approach.